### ISSUE TITLE
Remove references to `syscall` package

### ISSUE THREAD

**Issue by rolinh** [cilium/cilium#10116]
The [`syscall`](https://golang.org/pkg/syscall/) package is deprecated since Go 1.4:
> Deprecated: this package is locked down. Callers should use the corresponding package in the golang.org/x/sys repository instead. That is also where updates required by new systems or versions should be applied. See https://golang.org/s/go1.4-syscall for more information.

References to the `syscall` package in the codebase should be cleaned up as it is, in effect, not updated [_"not even to keep pace with changes in operating systems it references"_](https://golang.org/s/go1.4-syscall).

**Comment by Ropes** [comment]
Hey @Rolinh, I'm still working on understanding the architecture of Cilium, but interested in the project, and got `make` running successfully! 

Happy to take on this find/replace task following the [go1.4-syscall](https://golang.org/s/go1.4-syscall) directives. Any further instructions/guidance?

**Comment by Rolinh** [comment]
Hey @Ropes, thanks for contributing!

> got make running successfully!

You may want to take a look at the [contributor's guide](https://docs.cilium.io/en/stable/contributing/contributing/) which has instructions to bring up a vagrant VM for dev purposes.

> Any further instructions/guidance?

Nothing that I can think of as of now.

**Comment by Ropes** [comment]
@Rolinh thanks! I had a look at the contributor's guide, I'll get this branch running on Vagrant before PR.

### Complication

I discovered a use of [`sycall.RawConn`](https://golang.org/pkg/syscall/#RawConn) in `pkg/fqdn/dnsproxy`, [udp.go](https://github.com/cilium/cilium/blob/master/pkg/fqdn/dnsproxy/udp.go#L95-L105). This is a tight coupling of `syscall` and the `net` package's [`ListenConfig.ListenPackage(...)`](https://golang.org/pkg/net/#ListenConfig.ListenPacket) implementation. There is no `unix` package equivalent, nor would the type system allow it.

Some design criteria of the DNS/UDP forwarding are written in [SetSocketOptions](https://github.com/cilium/cilium/blob/master/pkg/fqdn/dnsproxy/udp.go#L131-L140):
```
	// Set up the raw socket for sending responses.
	// - Must use a raw UDP socket for sending responses so that we can send
	//   from a specific port without binding to it.
	// - The raw UDP socket must be bound to a specific IP address to prevent
	//   it receiving ALL UDP packets on the host.
	// - We use oob data to override the source IP address when sending
	// - Must use separate sockets for IPv4/IPv6, as sending to a v6-mapped
	//   v4 address from a socket bound to "::1" does not work due to kernel
	//   checking that a route exists from the source address before
	//   the source address is replaced with the (transparently) changed one
```
I don't posses a deep understanding of the UDP forwarding. However looking at the functionality of the package, and the design criteria above, I'm unsure if granular control of the `net.ListenConfig`->`ListenPacket()` can be achieved with the more connection based `net` package listener implementations. It might be possible, but it will be a more significant change. I can look into it, but I'd like feedback to determine if that's in scope.

### Decision

* Should the DNS proxy UDP listener be refactored to remove one use of `syscall.RawConn`? 
* I've [replaced](https://github.com/Ropes/cilium/commit/f414f188f8a2c748a4f06526136e9619e5df8cad) all other uses of `syscall` with `x/sys/unix`; going with the assumption that executing on Windows environments is not a project goal? Didn't see anything in the docs about running Windows k8s clusters so figured I could skip handling that. :D

**Comment by Rolinh** [comment]
> Should the DNS proxy UDP listener be refactored to remove one use of syscall.RawConn?

Not sure. /cc @jrajahalme 

> I've replaced all other uses of syscall with x/sys/unix; 

I think this is fine; Windows support would require a lot of work anyway.

**Comment by jrajahalme** [comment]
`syscall.RawConn` is an interface that is part of the `net` API. As long as there is no replacement for it within `net.ListenConfig` IMO we have to keep using it. The interface itself has no references to the rest of the `syscall` package, it just happens to be defined there. Even the implementation of it in https://golang.org/src/net/rawconn.go only references syscall for `syscall.EINVAL`. Given this I see no immediate reason for `net` to change away from using `syscall.RawConn`.

All other references to `syscall` in `pkg/fqdn/dnsproxy/udp.go` should be trivially replaceable with `unix`.

**Comment by Ropes** [comment]
Alright, I opened a PR to get feedback from the CI bots. Looks like Hound is not happy with some of the function return parameters in the files I updated. 

![Screenshot from 2020-02-12 00-02-06](https://user-images.githubusercontent.com/489062/74314876-1afde200-4d2b-11ea-8ce9-e712024dc1e5.png)

Are these sort of formatting critiques blocking? If you'd like me to make changes I'm open to do it, but also don't want to end up in a mess of Yak fur if that's counter-productive..

I also booted up a local testing environment, which built(!), but have yet to run tests... Wanted to get this posted for those in EU.
